[mlir][Transforms] Dialect conversion: Align handling of dropped values

matthias-springer opened the pull request:
Handle dropped block arguments and dropped op results in the same way: build a source materialization (that may fold away if unused). This simplifies the code base a bit and makes it possible to merge `legalizeConvertedArgumentTypes` and `legalizeConvertedOpResultTypes` in a future commit. These two functions are almost doing the same thing now.

As a side effect, this commit also changes the dialect conversion such that temporary circular cast ops are no longer generated. (There was a workaround in #107109 that can now be removed again.) Example:
```
%0 = "builtin.unrealized_conversion_cast"(%1) : (!a) -> !b
%1 = "builtin.unrealized_conversion_cast"(%0) : (!b) -> !a
// No further uses of %0, %1.
```

This happened when:
1. An op was erased. (No replacement values provided.)
2. A conversion pattern for another op builds a replacement value for the erased op's results (first cast op) during `remapValues`, but that SSA value is not used during the pattern application.
3. During the finalization phase, `legalizeConvertedOpResultTypes` thinks that the erased op is alive because of the cast op that was built in Step 2. It builds a cast from that replacement value to the original type.
4. During the commit phase, all uses of the original op are replaced with the casted value produced in Step 3. We have generated circular IR.

This problem can be avoided by making sure that source materializations are generated for all dropped results. This ensures that we always have some replacement SSA value in the mapping. Previously, we sometimes had a value mapped and sometimes not. (No more special casing is needed anymore to distinguish between "value dropped" or "value replaced with SSA value".)

matthias-springer committed:
[mlir][Transforms] Dialect conversion: Align handling of dropped values

Handle dropped block arguments and dropped op results in the same way: build a source materialization (that may fold away if unused). This simplifies the code base a bit and makes it possible to merge `legalizeConvertedArgumentTypes` and `legalizeConvertedOpResultTypes` in a future commit. These two functions are almost doing the same thing now.

This commit also fixes a bug where circular materializations were built, e.g.:
```
%0 = "builtin.unrealized_conversion_cast"(%1) : (!a) -> !b
%1 = "builtin.unrealized_conversion_cast"(%0) : (!b) -> !a
// No further uses of %0, %1.
```

This happened when:
1. An op was erased. (No replacement values provided.)
2. A conversion pattern for another op builds a replacement value (first cast op) during `remapValues`, but that SSA value is not used during the pattern application.
3. During the finalization phase, `legalizeConvertedOpResultTypes` thinks that the erased op is alive because of the cast op that was built in Step 2. It builds a cast from that replacement value to the original type.
4. During the commit phase, all uses of the original op are repalced with the casted value produced in Step 3. We have generated circular IR.

matthias-springer committed:
Remove `changedResults`

zero9178 reviewed (commented):
Is my intuition correct that this is rather difficult to test until the other PR relands but could then be tested by checking for a `() -> i32` `unrealized_conversion_cast` remaining in the IR (in the scenario showin in `@remove_all_ops`)?

Or could this maybe be tested right now be checking scenarios where previously two materializations would be created but now only one is?

matthias-springer commented:
@zero9178 I added a test case for the circular casts to #107109. I added an extra piece of code there to handle the circular casts. This piece of code can probably be deleted again with this PR.

I'd like to land #106760 first because I already ran it by Google and IREE folks and it should be fine (they have much more code that stress-tests the dialect conversion). #106760 is also needed more urgently by CIRCT people in their LLVM integration (see comments here: https://github.com/llvm/llvm-project/pull/106778#issuecomment-2326430255). I'd like to keep a few days between each merged dialect conversion commit so that it will be easier to address issues (if any). So, landing this PR not before next week or so...

zero9178 reviewed (approved):
Amazing, thank you so much!
